src/object.c

```diff
@@ -211,6 +211,361 @@ robj *dupStringObject(const robj *o) {
+            }
+            hashTypeCurrentFromZiplist(hi, OBJ_HASH_VALUE, &vstr, &vlen, &vll);
+            if (vstr) {
+                hvalue = sdsnewlen((const char *)vstr, vlen);
+            } else {
+                ll2string(buf, 64, vll);
+                hvalue = sdsnewlen((const char *)buf, strlen(buf));
+            }
+
+            /* Add a key-value pair to a new hash object. */
+            hashTypeSet(hobj, hkey, hvalue, HASH_SET_COPY);
+            sdsfree(hkey);
+            sdsfree(hvalue);
+        } else if (hi->encoding == OBJ_ENCODING_HT) {
+            sds hkey, hvalue;
+            /* Extract a key-value pair from an original hash object.*/
+            hkey = hashTypeCurrentFromHashTable(hi, OBJ_HASH_KEY);
+            hvalue = hashTypeCurrentFromHashTable(hi, OBJ_HASH_VALUE);
+
+            /* Add a key-value pair to a new hash object. */
+            hashTypeSet(hobj, hkey, hvalue, HASH_SET_COPY);
+            sdsfree(hkey);
+            sdsfree(hvalue);
+        } else {
+            serverPanic("Unknown hash encoding");
+        }
+    }
+    hashTypeReleaseIterator(hi);
+    return hobj;
+}
+
+/* This is a helper function for the COPY command.
+ * Duplicate a Stream object, with the guarantee that the returned object
+ * has the same encoding as the original one.
+ * 
+ * The resulting object always has refcount set to 1 */
+robj *dupStreamObject(robj *o) {
+    robj *sobj;
+    
+    serverAssert(o->type == OBJ_STREAM);
+
+    switch (o->encoding) {
+        case OBJ_ENCODING_STREAM:
+            sobj = createStreamObject();
+            break;
+        default:
+            serverPanic("Wrong encoding.");
+            break;
+    }
+
+    streamID id;
+    stream *s;
+    stream *new_s;
+    streamID startid;
+    streamID endid;
+    int64_t numfields;
+    int rev = 0;
+
+    startid.ms = startid.seq = 0;
+    endid.ms = endid.seq = UINT64_MAX;
+    s = o->ptr;
+    new_s = sobj->ptr;
+
+    streamIterator si;
+    streamIteratorStart(&si, s, &startid, &endid, rev);
+    while (streamIteratorGetID(&si, &id, &numfields)) {
+        /* Extract field-value pairs from an original stream object
+         * and, add these to a new stream object. */
+        robj **argv;
+        argv = zmalloc(sizeof(robj *) * numfields * 2);
+        for (int j = 0; j < numfields; j++) {
+            unsigned char *key, *value;
+            int64_t key_len, value_len;
+
+            streamIteratorGetField(&si, &key, &value, &key_len, &value_len);
+            argv[j * 2] = createObject(OBJ_STRING, sdsnewlen((const char *)key, key_len));
+            argv[j * 2 + 1] = createObject(OBJ_STRING, sdsnewlen((const char *)value, value_len));
+        }
+        streamAppendItem(new_s, argv, numfields, &id, &id);
+        zfree(argv);
+    }
+    streamIteratorStop(&si);
+
+    if (s->cgroups == NULL) {
+        /*Nothing to do*/
+    } else {
+        /* Consumer Groups */
+        raxIterator ri_cgroups;
```

Comment:
@guybe7 Do you agree that it makes sense to copy all of the consumer group information when we copy the stream. It seems weird to me, but it is metadata on the stream but also isn't directly data about the stream. Do you have any thoughts?

Reply:
@madolson yes, i think CGs and consumers should be copied as well... as i see it, COPY should provide a perfect copy of the key, i.e. executing the same command on both keys should have identical result